raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java

```diff
@@ -779,39 +781,43 @@ private VoteResponseData handleVoteRequest(
         VoteRequestData.PartitionData partitionRequest =
             request.topics().get(0).partitions().get(0);
 
-        int candidateId = partitionRequest.candidateId();
-        int candidateEpoch = partitionRequest.candidateEpoch();
+        int replicaId = partitionRequest.replicaId();
+        int replicaEpoch = partitionRequest.replicaEpoch();
+        boolean preVote = partitionRequest.preVote();
 
         int lastEpoch = partitionRequest.lastOffsetEpoch();
         long lastEpochEndOffset = partitionRequest.lastOffset();
-        if (lastEpochEndOffset < 0 || lastEpoch < 0 || lastEpoch >= candidateEpoch) {
+        boolean isIllegalEpoch = preVote ? lastEpoch > replicaEpoch : lastEpoch >= replicaEpoch;
```

Comment:
I see. Let's write a comment explaining this check.

This is my understanding. When the replica is sending a normal vote request the replica's epoch must be greater than the largest epoch on their log. This is true because the candidate replica will always request a vote for an epoch for which there is no previously known leader, or record in the log.

When the pre-vote flag is true then there may be a known leader for the last epoch so there may be a record in the log for that epoch.

Both of these cases should be rare and indicate a programming error in the remote replica. Let's log an INFO message with all of the relevant information when this condition is true.

Reply:
thank you!